client/internal/retry/do.go

```diff
@@ -83,62 +93,59 @@ func NewOptions(logger *log.Logger, timeout time.Duration, delay time.Duration,
 }
 
 // Do run retry function until response of request satisfy check function, or ends early according to configuration.
-func Do(retryFunc Func, opt Options) error {
-
-	startTime := time.Now()
-	endTime := startTime.Add(opt.Timeout)
-	timeout := func() bool {
-		return time.Now().After(endTime)
-	}
-	willTimeoutAfterDelay := func() bool {
-		return time.Now().Add(opt.Delay).After(endTime)
-	}
-	accumulatedErrs := make([]error, 0, opt.Retries+1)
-
-	// check if it is already timeout
-	if timeout() {
-		return fmt.Errorf("retry func timeout before any attempt")
+func Do(ctx context.Context, retryFunc Func, opt Options) error {
+	// set up context
+	ctxToUse := ctx
+	if opt.Timeout > 0 {
+		var cancel context.CancelFunc
+		ctxToUse, cancel = context.WithTimeout(ctx, opt.Timeout)
+		defer cancel()
 	}
+	return doInternal(ctxToUse, retryFunc, opt)
```

Comment:
I think that should be in the "if" scope, no?
maybe you can validate `opt` at the beginning of the method and avoid all the further checks

Reply:
The `if` scope overwrites context according to retry options passed in. The `defer` keyword is not executed after exiting `if` scope, it will only run after exit of the function scope, so I think it is fine to be outside of the scope.

Actually there is no invalid value for retry options because:
- negative timeout is treated as no time out
- negative delay is treated as no delay
The reason for this convention is the golang lib also have similar convention